Source/Core/Core/AchievementManager.cpp

```diff
@@ -1341,17 +1335,44 @@ void AchievementManager::ActivateDeactivateAchievement(AchievementId id, bool en
     rc_runtime_deactivate_achievement(&m_runtime, id);
 }
 
-void AchievementManager::GenerateRichPresence()
+void AchievementManager::DoPeriodically()
 {
-  Core::RunAsCPUThread([&] {
-    std::lock_guard lg{m_lock};
-    rc_runtime_get_richpresence(
-        &m_runtime, m_rich_presence.data(), RP_SIZE,
-        [](unsigned address, unsigned num_bytes, void* ud) {
-          return static_cast<AchievementManager*>(ud)->MemoryPeeker(address, num_bytes, ud);
-        },
-        this, nullptr);
-  });
+  if (!m_is_game_loaded)
+    return;
+  auto current_time = std::chrono::steady_clock::now();
+  if (current_time - m_last_rp_time > std::chrono::seconds{10})
+  {
+    if (Config::Get(Config::RA_RICH_PRESENCE_ENABLED))
+    {
+      Core::RunAsCPUThread([&] {
+        std::lock_guard lg{m_lock};
+        rc_runtime_get_richpresence(
+            &m_runtime, m_rich_presence.data(), RP_SIZE,
+            [](unsigned address, unsigned num_bytes, void* ud) {
+              return static_cast<AchievementManager*>(ud)->MemoryPeeker(address, num_bytes, ud);
+            },
+            this, nullptr);
+      });
+    }
+    m_last_rp_time = current_time;
+    m_update_callback();
+
+    if (current_time - m_last_ping_time > std::chrono::minutes{2})
```

Comment:
Not entirely sure what rich presence is but is there a reason the get and the ping are in the same block?